### PR TITLE
Itk5 threading

### PR DESCRIPTION
**Pull request by thewtex**
@dzenanz please take a look. Calls for parallel operations are **sooo** much cleaner!

### CODE REVIEW

File: include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx [hunk 1]
```diff
@@ -344,19 +360,28 @@ BayesianRegularizationDisplacementCalculator< TMetricImage, TDisplacementImage >
...
     for( fit = faceList.begin(); fit != faceList.end(); ++fit )
       {
-      ImpartLikelihoodThreadStruct str;
-      str.self = this;
-      this->m_DisplacementImage->SetRequestedRegion( *fit );
-      this->m_Threader->SetSingleMethod(
-        this->ImpartLikelihoodThreaderCallback, &str );
-      this->m_Threader->SingleMethodExecute();
+      // TODO: FIX-ME: hangs if this is not set
+      this->m_MultiThreader->SetNumberOfWorkUnits( 2 );
```
Comment (thewtex) [inline include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: @dzenanz can you please look into this? I observe a hang with the `itk::PoolMultiThreader`. I suspect that the region is smaller than the number of work units requested, but the threading backends should be able to handle this condition gracefully.

**Comment by thewtex** [comment]
@dzenanz :+1: Thanks, Dzenan.

I like the workaround! I will test it locally, then consider upstreaming a re-usable function in ITK.

File: include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx (hunk 1 above)
Comment (dzenanz) [reply include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: It doesn't hang on my Windows computer even if I remove `this->m_MultiThreader->SetNumberOfWorkUnits( 2 );` I will test on Linux too.

Comment (dzenanz) [reply include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: On Linux I had [this](http://review.source.kitware.com/#/c/23576/) branch built, and with it there is no hang, whether using Pool or TBB. I didn't test with master on Linux.

Comment (thewtex) [reply include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: Likely related -- this was testing on a 12-core system.

However, I tried with the `PoolMultiThreader` refactoring patch, and it passed! :fireworks:

Comment (dzenanz) [reply include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: Excellent! Once PoolMultiThreader refactoring gets merged, we can remove this workaround and merge this and/or #81.

Comment (dzenanz) [reply include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: PoolMultiThreader refactoring has been merged, so this is good to go I suppose.

Comment (dzenanz) [reply include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx]: Don't forget to remove this.

**Comment by thewtex** [comment]
Moved the `SetNumberOfWorkUnits` patch to #82 so we can keep the CI green while iterating on other tasks.

**Comment by thewtex** [comment]
The Python builds still have issues, but they should be addressed with ITK 5 Beta 1. Merging to forge ahead with other tasks.

**Pull request merged**